### ISSUE TITLE
New feature: `split`

### ISSUE THREAD

**Issue by noahlaux** [taskcluster/json-e#368]
# Split

This issue suggest to create a builtin `split` functionality as available in many languages, and is not possible using any other (or combination of) exciting builtins.

Splitting on strings with a delimiter is the most obvious, but this suggestion also give a take on splitting list values and further justify the potential use cases.

Finally there's a [specification](#specification) for the suggestion, which I hope to be more or less feature complete.

Have a look and give some feedback, I for one could use this functionality, and I'm willing to make a PR with whatever we come up with. The only thing I'd like to add would maybe to be able to split on regex as well, but I couldn't come up with a use case where it would make sense.

## Splitting strings

```yaml
context:
  topic: boston/thermometer/723e11c9-75d3-4cef-a921-94304ddf6b1b/temperature
template:
  $let:
    segments: {$eval: split(topic, '/')}
  in:
    city: segments[0]
    type: segments[1]
    uuid: segments[2]
    property: segments[3]
result:
  city: boston
  type: thermometer
  uuid: 723e11c9-75d3-4cef-a921-94304ddf6b1b
  property: temperature
```

```yaml
context:
  time: 9
template:
  $let:
    periods:
      0-5: night
      6-9: morning
      10-12: noon
      13-17: afternoon
      18-23: evening
  in:
    $map: {$eval: periods}
    each(value, key):
      $let:
        range: {$eval: split(key, '-')}
      in:
        $if: "time >= range[0] && time <= range[1]"
        then:
          message: ${time} is in the ${value}
result:
  message: 9 is in the morning
```

## Splitting arrays

```yaml
context:
  Uint8Array: [109, 101, 115, 115, 97, 103, 101, 32, 49, 10, 109, 101, 115, 115, 97, 103, 101, 32, 50]
template:
  $map: {$eval: split(Uint8Array, 10)} # 10 is a newline character used as delimiter in byte stream
  each(section): ${decode(section)}
result:
  - message 1
  - message 2
```

## Crazy

Granted, this example went a little into the future, but it shows how split could be used in a nice
and concise way though 2 additional features are needed in order to achieve this:

- Declarations of tuples from array input would make splitting more sexy _(the `$let` section with `platform` and `version`)_
- An implementation of a string formatter (fx index based on arguments as shown)

```yaml
context:
  dependencies:
    - node:5.2
    - go:1.3
template:
  $let:
    base:
      node: https://nodejs.org/dist/v{0}/node-v{0}-linux-armv7l.tar.xz
      go: https://golang.org/dl/go{0}.linux-amd64.tar.gz
  in:
    $map: {$eval: dependencies}
    each(dependency):
      $let:
        [platform, version]: {$eval: split(dependency, ':')}
      in: ${format(base[platform], version)}
result:
  - https://nodejs.org/dist/v5.2/node-v5.2-linux-armv7l.tar.xz
  - https://golang.org/dl/go1.3.linux-amd64.tar.gz
```

## Specification

```yaml
---
title: split string with delimiter
context: {}
template: {$eval: "split('left:right', ':')"}
result: ['left', 'right']
---
title: split string without delimiter
context: {}
template: {$eval: "split('splitme', '')"}
result: ['s', 'p', 'l', 'i', 't', 'm', 'e']
---
title: split with empty string
context: {}
template: {$eval: "split('', ':')"}
result: ['']
---
title: split string with no match
context: {}
template: {$eval: "split('no semicolon', ':')"}
result: ['no semicolon']
---
title: split with number delimiter
context: {}
template: {$eval: "split('left1right', 1)"}
result: ['left', 'right']
---
title: split arrays
context: {}
template: {$eval: "split([1, 2, 3, 4, 5, 6, 7], 4)"}
result: [[1, 2, 3], [5, 6, 7]]
---
title: split arrays with no match
context: {}
template: {$eval: "split([1, 2, 3, 4, 5, 6, 7], 9)"}
result: [[1, 2, 3, 4, 5, 6, 7]]
---
title: split arrays with empty array
context: {}
template: {$eval: "split([], 1)"}
result: []
---
title: split arrays starting with delimiter
context: {}
template: {$eval: "split([4, 1, 2, 3, 4, 5], 4)"}
result: [[], [1, 2, 3], [5]]
---
title: split arrays with no delimiter
context: {}
template: {$eval: "split([1, 2, 3, 4])"}
result: [[1], [2], [3], [4]]
---
title: split arrays with wrong type
context: {}
template: {$eval: "split(false)"}
error: split(value) only works on strings, numbers and arrays
---
title: split error
context: {}
template: {$eval: "split({})"}
error: 'BuiltinError: invalid arguments to builtin: split'
---
```

**Comment by djmitche** [comment]
Thank you for the thorough post!

After some thought, my reactions are pretty concise:

 * `split` on strings seems like a great addition -- let's do it
 * `split` with regular expressions is also attractive, but we have avoided embedding regular expressions in JSON-e since each base-language's implementation of regexps is a little different.  See #331.
 * If we have `split`, we should probably also have `join`.
 * I'm not sold on splitting arrays.  The use-case presented, splitting a byte string, is legitimate, but I suspect handling byte strings with JSON-e is unusual.  I'd like to see a broader set of use-cases for this functionality, and perhaps input from other users on the utility of this design versus alternatives, before implementing it.
 * Destructuring assignment (`{"let": {"[a, b]": ".."}}`) is appealing, but it's a very "general" kind of syntax that introduces a lot of corner cases that would need to be considered carefully.  It is more efficient than calculating the same intermediate value multiple times, but #263 would help with this:
```yaml
$let:
  - intermediate: ".."
  - a: intermediate[0]
  - b: intermediate[1]
```
 * We already have a kind of string formatting with `"${..}"`, so I'm not sure what this additional form would offer.
 * Both destructuring and string formatting should, if proposed, get their own issue for discussion -- let's focus on split (and join) here.

**Comment by noahlaux** [comment]
I added PRs for the [split](#369) and [join](#371) features

**Comment by djmitche** [comment]
Released in v4.3.0
